core/types/gen_tx_json.go

```diff
@@ -11,91 +11,181 @@ import (
 	"github.com/ethereum/go-ethereum/common/hexutil"
 )
 
-var _ = (*txdataMarshaling)(nil)
-
-// MarshalJSON marshals as JSON.
-func (t txdata) MarshalJSON() ([]byte, error) {
+// MarshalJSONWithHash marshals as JSON with a hash.
+func (t *Transaction) MarshalJSON() ([]byte, error) {
 	type txdata struct {
+		Type         hexutil.Uint64  `json:"type"     rlp:"-"`
+		Chain        *hexutil.Big    `json:"chainId"  rlp:"-"`
 		AccountNonce hexutil.Uint64  `json:"nonce"    gencodec:"required"`
 		Price        *hexutil.Big    `json:"gasPrice" gencodec:"required"`
 		GasLimit     hexutil.Uint64  `json:"gas"      gencodec:"required"`
 		Recipient    *common.Address `json:"to"       rlp:"nil"`
 		Amount       *hexutil.Big    `json:"value"    gencodec:"required"`
 		Payload      hexutil.Bytes   `json:"input"    gencodec:"required"`
+		AccessList   *AccessList     `json:"accessList" rlp:"-"`
 		V            *hexutil.Big    `json:"v" gencodec:"required"`
 		R            *hexutil.Big    `json:"r" gencodec:"required"`
 		S            *hexutil.Big    `json:"s" gencodec:"required"`
 		Hash         *common.Hash    `json:"hash" rlp:"-"`
 	}
 	var enc txdata
-	enc.AccountNonce = hexutil.Uint64(t.AccountNonce)
-	enc.Price = (*hexutil.Big)(t.Price)
-	enc.GasLimit = hexutil.Uint64(t.GasLimit)
-	enc.Recipient = t.Recipient
-	enc.Amount = (*hexutil.Big)(t.Amount)
-	enc.Payload = t.Payload
-	enc.V = (*hexutil.Big)(t.V)
-	enc.R = (*hexutil.Big)(t.R)
-	enc.S = (*hexutil.Big)(t.S)
-	enc.Hash = t.Hash
+	enc.AccountNonce = hexutil.Uint64(t.Nonce())
+	enc.Price = (*hexutil.Big)(t.GasPrice())
+	enc.GasLimit = hexutil.Uint64(t.Gas())
+	enc.Recipient = t.To()
+	enc.Amount = (*hexutil.Big)(t.Value())
+	enc.Payload = t.Data()
+	v, r, s := t.RawSignatureValues()
+	enc.V = (*hexutil.Big)(v)
+	enc.R = (*hexutil.Big)(r)
+	enc.S = (*hexutil.Big)(s)
+	hash := t.Hash()
+	enc.Hash = &hash
+	if t.Type() == AccessListTxId {
+		enc.Type = hexutil.Uint64(t.Type())
+		enc.Chain = (*hexutil.Big)(t.ChainId())
+		enc.AccessList = t.AccessList()
+	}
 	return json.Marshal(&enc)
 }
 
 // UnmarshalJSON unmarshals from JSON.
-func (t *txdata) UnmarshalJSON(input []byte) error {
+func (t *Transaction) UnmarshalJSON(input []byte) error {
+	type typ struct {
+		Type *hexutil.Uint64 `json:"type" rlp:"-"`
+	}
```

Comment:
Why did this wind up in it's own struct? Being of pointer type, seems like it could be read simultaneously as everything else, regardless of whether it's present or not?

Reply:
I'm not sure now, it looks like it would be fine to be in the `txdata` struct.

Reply:
Yeah, I didn't mention it, but I already changed that, it worked well

Reply:
I noticed when I went to change it :)